### ISSUE TITLE
Panic on missing key in JSON facts 

### ISSUE THREAD

**Issue by theonedom** [hyperjumptech/grule-rule-engine#270]
**Describe the bug**
I am seeing a panic occur during `execute` in case a key is not provided in the JSON fact structure.

```
Exception has occurred: panic
"reflect: call of reflect.Value.Elem on zero Value"
Stack:
	 2  0x00000000004c4fe6 in reflect.Value.Elem
	     at /usr/lib/go/src/reflect/value.go:843
	 3  0x0000000000a7d97b in github.com/hyperjumptech/grule-rule-engine/model.(*JSONValueNode).GetObjectValueByField
	     at /home/<username>/go/pkg/mod/github.com/hyperjumptech/grule-rule-engine@v1.10.2/model/JsonDataAccessLayer.go:198
	 4  0x0000000000a80ccf in github.com/hyperjumptech/grule-rule-engine/model.(*JSONValueNode).GetChildNodeByField
	     at /home/<username>/go/pkg/mod/github.com/hyperjumptech/grule-rule-engine@v1.10.2/model/JsonDataAccessLayer.go:331
```
**To Reproduce**
```
const (
	input_rule = `
	rule TestRule "" {
		when
			R.Result == 'NoResult' &&
			inputs.i_am_missing == 'abc' &&
                        inputs.name.first == 'john'
		then
			R.Result = "ok";
	}
	`
)

func main() {

	oresult := &ObjectResult{
		Result: "NoResult",
	}

	// build rules
	lib := ast.NewKnowledgeLibrary()
	rb := builder.NewRuleBuilder(lib)
	err := rb.BuildRuleFromResource("Test", "0.0.1", pkg.NewBytesResource([]byte(input_rule)))

	// 	add JSON fact
	json := []byte(`{"blabla":"bla","name":{"first":"john","last":"doe"}}`)
	kb := lib.NewKnowledgeBaseInstance("Test", "0.0.1")
	dcx := ast.NewDataContext()

	err = dcx.Add("R", oresult)
	err = dcx.AddJSON("inputs", json)
	if err != nil {
		fmt.Println(err.Error())
	}

	// results in panic
	engine.NewGruleEngine().Execute(dcx, kb)

}
```

**Expected behavior**
As presence of individual keys/entries in the struct may not be validated for JSON facts, `JSONDataAccessLayer` should handle gracefully handle cases like this. Ideally:
- return an empty data point for use in the rule (which will then fail).
- logging or optionally returning an error.

A naive approach (always returning a string) could be to modify `GetObjectValueByField` in `JsonDataAccessLayer.go`:
```
- return tmap.Elem(), nil
+ if !tmap.IsValid() {
+    return reflect.ValueOf(""), nil
+} else {
+	return tmap.Elem(), nil
+}
```

**Additional context**
We have wrapped Grule in a small microservice. Rules and facts are submitted by an external service; meaning there is no validation that the JSON contains all expected fields prior to rule execution (unless the downstream service implements it).

**Comment by newm4n** [comment]
Hi @theonedom 

You are correct, in JSON, when you evaluate JSON values and the key do not exist, the evaluation should fail naturally.
The fix is rather simple, make sure that the evaluated value is checked for `nil`/`zero`/`reflect.valueOf(nil)` and  return a simple error if its `nil`/`zero`.

The commit  https://github.com/hyperjumptech/grule-rule-engine/pull/271 fixes this. Also added your reproduce snippet as a new test.

**Comment by newm4n** [comment]
Its now fixed in v1.10.4 . https://github.com/hyperjumptech/grule-rule-engine/releases/tag/v1.10.4
